onnxruntime/core/providers/webnn/builders/impl/tile_op_builder.cc

```diff
@@ -0,0 +1,98 @@
+ private:
+  bool IsOpSupportedImpl(const InitializedTensorSet& initializers, const Node& node,
+                         const WebnnDeviceType /* device_type */, const logging::Logger& logger) const override;
+};
+
+// Add operator related.
+
+void TileOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const {
+  model_builder.AddInitializerToSkip(node.InputDefs()[1]->Name());
+}
+
+Status TileOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                            const Node& node,
+                                            const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+  const auto& initializers(model_builder.GetInitializerTensors());
+  const auto& repetitions_initializer = *initializers.at(input_defs[1]->Name());
+  const int64_t* raw_repetitions_data = repetitions_initializer.int64_data().empty()
+                                            ? reinterpret_cast<const int64_t*>(repetitions_initializer.raw_data().data())
+                                            : repetitions_initializer.int64_data().data();
+  const auto size = repetitions_initializer.dims()[0];
+  TensorShapeVector repetitions_data{raw_repetitions_data, raw_repetitions_data + size};
+  emscripten::val input = model_builder.GetOperand(input_defs[0]->Name());
+  std::vector<int32_t> repetitions;
+  std::transform(repetitions_data.cbegin(), repetitions_data.cend(),
+                 std::back_inserter(repetitions),
+                 [](int64_t repetition) -> uint32_t { return SafeInt<int32_t>(repetition); });
+
+  emscripten::val options = emscripten::val::object();
+  options.set("label", node.Name());
+  emscripten::val output = model_builder.GetBuilder().call<emscripten::val>("tile",
+                                                                            input,
+                                                                            emscripten::val::array(repetitions),
+                                                                            options);
+  model_builder.AddOperand(node.OutputDefs()[0]->Name(), std::move(output));
+  return Status::OK();
+}
+
+// Operator support related.
+
+bool TileOpBuilder::IsOpSupportedImpl(const InitializedTensorSet& initializers,
+                                      const Node& node,
+                                      const WebnnDeviceType /* device_type */,
+                                      const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+  const auto& repetitions_name = input_defs[1]->Name();
+  if (!Contains(initializers, repetitions_name)) {
+    LOGS(logger, VERBOSE) << "Repetitions of tile must be a constant initializer";
+    return false;
+  }
+
+  std::vector<int64_t> input_shape;
+  if (!GetShape(*input_defs[0], input_shape, logger))
+    return false;
+
+  if (input_shape.empty()) {
+    LOGS(logger, VERBOSE) << "Tile does not support empty input shape";
```

Comment:
Interestingly ORT's DML EP actually runs fine with a 0D scalar, as the input dimensions rank will be 0, and the repetitions will also be zero length:

![image](https://github.com/user-attachments/assets/81731126-6147-4804-941f-66ff923400c3)

```
    {
      "graph_name": "Tile 0D float32",
      "op_type": "Tile",
      "input": 1,
      "repeats": [],
      "output": 1,
      "T": "float32"
    },
```

It's not a particularly *interesting* case, because it's just a nop, but it executes okay.

Reply:
WebNN EP doesn't support 0D input tensor currently, but chromium implementation allows 0d input tensor, right?

Reply:
WebNN EP supports 0D input, it is a scalar.

Reply:
Oh, some operators like `expand` and `reshape` don't support empty input dimensions.

Reply:
Right, this depends on how the WebNN spec states.

Reply:
> Oh, some operators like `expand` and `reshape` don't support empty input dimensions.

Hmm, sounds like a bug then, as a `reshape` from a scalar to a scalar should be valid (both input and output dimensions have the same rank), and for `expand`, a scalar with an empty shape would be a nop, but I don't think it should be an error.

Reply:
Do you mean WebNN EP should support the a scalar to a scalar `nop`  for both `expand` and `reshape`? In chromium dml implementation, we didn't handle a scalar to a scalar `nop`  for both `expand` and `reshape`.

Reply:
> Do you mean WebNN EP should support the a scalar to a scalar `nop` for both `expand` and `reshape`?

✅

>  In chromium dml implementation, we didn't handle a scalar to a scalar `nop` for both `expand` and `reshape`.

Then we should add some test cases for that case too 😉.

Reply:
Open issues https://issues.chromium.org/issues/369432531 and https://github.com/microsoft/onnxruntime/issues/22215 to track.